codegen-v2/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+#[macro_use]
+extern crate serde;
+
+use grammar::{GHeaderFileItem, ParseTree};
+use reader::Reader;
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+};
+
+mod grammar;
+mod reader;
+#[cfg(test)]
+mod tests;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub enum Error {
+    Todo,
+    Eof,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct CHeaderDirectory {
+    map: HashMap<PathBuf, Vec<GHeaderFileItem>>,
+}
+
+pub fn parse_dir(path: &Path, mut main_dir: CHeaderDirectory) -> Result<CHeaderDirectory> {
```

Comment:
I'm wondering, could `parse_dir` and `parse_file` take `main_dir: &mut CHeaderDirectory` argument and return `Result<()>` instead?
Just in my opinion, it's more clear 🙂